File: 0116.填充每个节点的下一个右侧节点指针/0116-填充每个节点的下一个右侧节点指针.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *left;
 *     struct Node *right;
 *     struct Node *next;
 * };
 */

void dfs(struct Node* root, struct Node* next) {
    if( !root ) {
        return;
    }

    root->next = next;
    dfs( root->left, root->right );
    dfs( root->right, root->next ? root->next->left : NULL );
}

struct Node* connect(struct Node* root) {
	dfs( root, NULL );
    return root;
}

File: 0116.填充每个节点的下一个右侧节点指针/0116-填充每个节点的下一个右侧节点指针.py
"""
# Definition for a Node.
class Node(object):
    def __init__(self, val=0, left=None, right=None, next=None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution(object):

    def connect(self, root):
        """
        :type root: Node
        :rtype: Node
        """
        self.dfs(root, None)
        return root

    def dfs( self, root, next ):
        if root == None :
            return 

        root.next = next
        self.dfs( root.left, root.right )
        self.dfs( root.right, None if root.next==None else root.next.left )